src/hotspot/share/gc/shenandoah/shenandoahFreeSet.cpp

```diff
@@ -392,61 +674,220 @@ void ShenandoahFreeSet::clear() {
+  size_t collector_empty_xfer = 0;
+  size_t collector_not_empty_xfer = 0;
+
+  // Process empty regions within the Collector free partition
+  if ((max_xfer_regions > 0) && (_partitions.leftmost_empty(Collector) <= _partitions.rightmost_empty(Collector))) {
+    ShenandoahHeapLocker locker(_heap->lock());
+    for (size_t idx = _partitions.leftmost_empty(Collector);
+         (max_xfer_regions > 0) && (idx <= _partitions.rightmost_empty(Collector)); idx++) {
+      if (_partitions.in_partition(idx, Collector) && can_allocate_from(idx)) {
+        _partitions.move_to_partition(idx, Mutator, region_size_bytes);
+        max_xfer_regions--;
+        collector_empty_xfer += region_size_bytes;
+      }
+    }
+  }
 
-      assert(!is_mutator_free(idx), "We are about to add it, it shouldn't be there already");
-      _mutator_free_bitmap.set_bit(idx);
+  // If there are any non-empty regions within Collector partition, we can also move them to the Mutator free partition
+  if ((max_xfer_regions > 0) && (_partitions.leftmost(Collector) <= _partitions.rightmost(Collector))) {
+    ShenandoahHeapLocker locker(_heap->lock());
+    for (size_t idx = _partitions.leftmost(Collector);
+         (max_xfer_regions > 0) && (idx <= _partitions.rightmost(Collector)); idx++) {
+      size_t alloc_capacity = this->alloc_capacity(idx);
+      if (_partitions.in_partition(idx, Collector) && (alloc_capacity > 0)) {
+        _partitions.move_to_partition(idx, Mutator, alloc_capacity);
+        max_xfer_regions--;
+        collector_not_empty_xfer += alloc_capacity;
+      }
     }
   }
 
-  // Evac reserve: reserve trailing space for evacuations
-  size_t to_reserve = _heap->max_capacity() / 100 * ShenandoahEvacReserve;
-  size_t reserved = 0;
+  size_t collector_xfer = collector_empty_xfer + collector_not_empty_xfer;
+  log_info(gc, free)("At start of update refs, moving " SIZE_FORMAT "%s to Mutator free partition from Collector Reserve",
+                     byte_size_in_proper_unit(collector_xfer), proper_unit_for_byte_size(collector_xfer));
+}
 
-  for (size_t idx = _heap->num_regions() - 1; idx > 0; idx--) {
-    if (reserved >= to_reserve) break;
 
-    ShenandoahHeapRegion* region = _heap->get_region(idx);
-    if (_mutator_free_bitmap.at(idx) && can_allocate_from(region)) {
-      _mutator_free_bitmap.clear_bit(idx);
-      _collector_free_bitmap.set_bit(idx);
-      size_t ac = alloc_capacity(region);
-      _capacity -= ac;
-      reserved += ac;
+// Overwrite arguments to represent the number of regions to be reclaimed from the cset
+void ShenandoahFreeSet::prepare_to_rebuild(size_t &cset_regions) {
+  shenandoah_assert_heaplocked();
+  // This resets all state information, removing all regions from all partitions.
+  clear();
+  log_debug(gc, free)("Rebuilding FreeSet");
+
+  // This places regions that have alloc_capacity into the mutator partition.
+  find_regions_with_alloc_capacity(cset_regions);
```

Comment:
We walk twice, first to figure out how much memory is available, and how many regions are completely empty.  This information eventually feeds into GenShen's transfer of regions between young-gen and old-gen.  There is less motivation for that distinction in single-generation Shenandoah because we do not need to make these informed transfers.

Reply:
But even before genshen changes, there were two walks through the regions.  This is because the rebuild wants to "optimize" the organization of the mutator free set and the collector free set.  Certain regions that may been in the mutator set during previous GC will be in the collector set during the next gc, and vice versa.  We strive to arrange that each free set is "tightly packed" over a subrange of the regions, with collector free set at the high end of memory and mutator set at the lower end of memory.  With GenShen integration, we will place the old collector set above the collector set.

Reply:
I suppose I'll need to look through this more carefully. In the case of single gen, it still sounded to me like the "clear" really doesn't accomplish anything other than taking stuff out of the free partitions and then the `find_..` sorts them into the new free partitions, and it looked like that could be accomplished by a single walk.

If GenShen then wants to break them into two sequences with some other step in between, may be one offers the three API's: one the single-gen optimized one that avoids two walks, and the two APIs `clear` and `find_...` separately for GenShen. But it sounds like you are saying that there is a _need_ for these two walks in the case of single gen as well. Let me discuss this with you offline so I understand this better as I am probably missing something crucial here. Thanks!

Reply:
In the most recent commit, I have merged the first two passes through the free set.  We no longer have a separate clear pass that precedes moving regions with available capacity into the mutator free set.